### PR TITLE
Add tests for comparators added in #6465

### PR DESCRIPTION
**Pull request by jostephd**
#6465 added two new comparators, ByGivenOrder and ByName.

This PR adds tests for them.

### CODE REVIEW

File: tests/src/comparators/test_byGivenOrder.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// #region unit tests
+SCENARIO( "Test basic ByGivenOrder functionality." , "[ByGivenOrder]" ) {
+	const std::vector<int> givenOrder = { 4, 2, 8, 6 };
+	std::vector<int> toSort = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+	const std::vector<int> expectedOrder = { 4, 2, 8, 6, 0, 1, 3, 5, 7, 9 };
+
+	std::sort(toSort.begin(), toSort.end(), ByGivenOrder<int>(givenOrder));
+	REQUIRE( toSort == expectedOrder );
+
+}
```
Comment (jostephd) [inline tests/src/comparators/test_byGivenOrder.cpp:31]: @tehhowch's reviewed on Discord: https://discord.com/channels/251118043411775489/285540320823869441/932688539118862516

Comment (jostephd) [reply tests/src/comparators/test_byGivenOrder.cpp]: Done in https://github.com/endless-sky/endless-sky/pull/6497/commits/93e792d33783141ec062c9d0c1971869402d0324

File: tests/src/comparators/test_byGivenOrder.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			std::sort(toSort.begin(), toSort.end(), comparator);
+			REQUIRE( toSort == expectedOrder );
+		}
+
+		THEN( "Unknown elements are sorted after known elements" ) {
+			std::vector<int> toSort = { 2, 4, 6, 8, 5, 1, 3 };
+			const std::vector<int> expectedOrder = { 4, 2, 8, 6, 1, 3, 5 };
+			std::sort(toSort.begin(), toSort.end(), comparator);
+			REQUIRE( toSort == expectedOrder );
+		}
```
Comment (petervdmeer) [inline tests/src/comparators/test_byGivenOrder.cpp:47]: The tests here look good to me.
Adding some tests with a givenOrder that has duplicates might be good for test-coverage (what if the number 4 appears on location 1 and on location 3?). And adding some tests where the toSort has duplicate numbers might also be good for test-coverage. But I consider none of those blocking for merging.

File: tests/src/comparators/test_byName.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	int Name() const { return name; };
+	int name;
+};
+// #endregion mock data
+
+
+
+// #region unit tests
+SCENARIO( "Test basic ByName functionality." , "[ByName]" ) {
+	NamedObject values[3] = { 10, 20, 30 };
```
Comment (petervdmeer) [inline tests/src/comparators/test_byName.cpp:35]: Doing a test with NamedObjects and then using numbers for them was a little confusing, but the code seems correct and the use of strings or numbers doesn't really matter for the test.

**Review by petervdmeer — APPROVED** [review]

File: tests/src/comparators/test_byGivenOrder.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/* test_byGivenOrder.cpp
+Copyright (c) 2022 by Michael Zahniser
```
Comment (petervdmeer) [inline tests/src/comparators/test_byGivenOrder.cpp:2]: You might still want to change the copyright.

File: tests/src/comparators/test_byName.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	REQUIRE( c(&values[0], &values[1]) == true );
+	REQUIRE( c(&values[0], &values[2]) == true );
+
+	REQUIRE( c(&values[1], &values[0]) == false );
+	REQUIRE( c(&values[1], &values[1]) == false );
+	REQUIRE( c(&values[1], &values[2]) == true );
+
+	REQUIRE( c(&values[2], &values[0]) == false );
+	REQUIRE( c(&values[2], &values[1]) == false );
+	REQUIRE( c(&values[2], &values[2]) == false );
```
Comment (tehhowch) [inline tests/src/comparators/test_byName.cpp:48]: These should use CHECK since we can safely execute them all. (REQUIRE should be used when subsequent tests would be ill-formed.)

File: tests/src/comparators/test_byGivenOrder.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+			std::vector<int> toSort = { 5, 1, 3 };
+			const std::vector<int> expectedOrder = { 1, 3, 5 };
+			std::sort(toSort.begin(), toSort.end(), comparator);
+			REQUIRE( toSort == expectedOrder );
+		}
+
+		THEN( "Unknown elements are sorted after known elements" ) {
+			std::vector<int> toSort = { 2, 4, 6, 8, 5, 1, 3 };
+			const std::vector<int> expectedOrder = { 4, 2, 8, 6, 1, 3, 5 };
+			std::sort(toSort.begin(), toSort.end(), comparator);
```
Comment (tehhowch) [inline tests/src/comparators/test_byGivenOrder.cpp:45]: This test would fail if any part of the comparator wasn't working, not just the bit its scenario identifies, which is why I suggested testing the comparator directly.

**Review by tehhowch — COMMENTED** [review]
Need to update the codeblocks test project

File: tests/src/comparators/test_byGivenOrder.cpp (hunk 2 above)
Comment (jostephd) [reply tests/src/comparators/test_byGivenOrder.cpp:68]: > Adding some tests with a givenOrder that has duplicates might be good for test-coverage (what if the number 4 appears on location 1 and on location 3?).

That's not a valid way to use ByGivenOrder. I guess it could raise an exception, or log an error, or something.

> And adding some tests where the toSort has duplicate numbers might also be good for test-coverage.

Added explicit tests of comparing something to itself.

File: tests/src/comparators/test_byName.cpp (hunk 3 above)
Comment (jostephd) [reply tests/src/comparators/test_byName.cpp:35]: Any changes needed? There's already a comment `// #region mock data`.

File: tests/src/comparators/test_byName.cpp (hunk 5 above)
Comment (jostephd) [reply tests/src/comparators/test_byName.cpp]: Done

File: tests/src/comparators/test_byGivenOrder.cpp (hunk 6 above)
Comment (jostephd) [reply tests/src/comparators/test_byGivenOrder.cpp]: How about now?

File: tests/src/comparators/test_byGivenOrder.cpp (hunk 4 above)
Comment (tehhowch) [reply tests/src/comparators/test_byGivenOrder.cpp:2]: @jostephd were you going to do this?

File: tests/src/comparators/test_byGivenOrder.cpp (hunk 2 above)
Comment (tehhowch) [reply tests/src/comparators/test_byGivenOrder.cpp:68]: I'm fine with the current behavior (any occurrence beyond the first is silently ignored).

It's possible we use this with user-controlled data where duplicates can occur, so throwing a runtime error would be undesirable.

File: tests/src/comparators/test_byGivenOrder.cpp (hunk 4 above)
Comment (jostephd) [reply tests/src/comparators/test_byGivenOrder.cpp:2]: I'm happy with that comment as is... does it need to be changed?

Comment (tehhowch) [reply tests/src/comparators/test_byGivenOrder.cpp:2]: From the 👍 I wasn't sure of your intent. Since you say it's fine as-is, fine by me.
